adcm_aio_client/core/config/_objects.py

```diff
@@ -0,0 +1,469 @@
+from copy import deepcopy
+from functools import partial
+from typing import Any, Callable, Coroutine, Protocol, Self, overload
+import json
+import asyncio
+
+from adcm_aio_client.core.config._operations import find_config_difference
+from adcm_aio_client.core.config.merge import apply_local_changes
+from adcm_aio_client.core.config.types import (
+    AnyParameterName,
+    ConfigData,
+    ConfigDifference,
+    ConfigSchema,
+    LevelNames,
+    LocalConfigs,
+    MergeStrategy,
+)
+from adcm_aio_client.core.errors import ConfigComparisonError, RequesterError
+from adcm_aio_client.core.types import AwareOfOwnPath, WithRequesterProperty
+
+
+class ConfigOwner(WithRequesterProperty, AwareOfOwnPath, Protocol): ...
+
+
+# Config Entries Wrappers
+
+
+class _ConfigWrapper:
+    __slots__ = ("_name", "_schema", "_data")
+
+    def __init__(
+        self: Self,
+        name: LevelNames,
+        data: ConfigData,
+        schema: ConfigSchema,
+    ) -> None:
+        self._name = name
+        self._schema = schema
+        self._data = data
+
+
+class _Group(_ConfigWrapper):
+    __slots__ = ("_name", "_schema", "_data", "_wrappers_cache")
+
+    def __init__(self: Self, name: LevelNames, data: ConfigData, schema: ConfigSchema) -> None:
+        super().__init__(name, data, schema)
+        self._wrappers_cache = {}
+
+    def _find_and_wrap_config_entry[ValueW: _ConfigWrapper, GroupW: _ConfigWrapper, AGroupW: _ConfigWrapper](
+        self: Self,
+        item: AnyParameterName | tuple[AnyParameterName, type[ValueW | GroupW | AGroupW]],
+        value_class: type[ValueW],
+        group_class: type[GroupW],
+        a_group_class: type[AGroupW],
+    ) -> ValueW | GroupW | AGroupW:
+        if isinstance(item, str):
+            name = item
+        else:
+            name, *_ = item
+
+        level_name = self._schema.get_level_name(group=self._name, display_name=name)
+        if level_name is None:
+            level_name = name
+
+        cached_wrapper = self._wrappers_cache.get(level_name)
+        if cached_wrapper:
+            return cached_wrapper
+
+        parameter_full_name = (*self._name, level_name)
+
+        class_ = value_class
+        if self._schema.is_group(parameter_full_name):
+            class_ = a_group_class if self._schema.is_activatable_group(parameter_full_name) else group_class
+
+        wrapper = class_(name=parameter_full_name, data=self._data, schema=self._schema)
+
+        self._wrappers_cache[level_name] = wrapper
+
+        return wrapper
+
+
+class Value[T](_ConfigWrapper):
```

Comment:
`ValueWrapper` is a bit long to use it for typehints, but if you insist, I can rename (see Desyncable also)

Parameter isn't the value, it's something like description of value (Value doesn't have name for now, for example)

Reply:
on second thought, Parameter may be a good term, let me think about it

Reply:
You are right, we can introduce term `Parameter`, because later we will most likely have schema associated with it alongside the value (description, type properties, etc.)

So naming can be like:

- Parameter
- ParameterGroup
- ActivatableParameterGroup (this one bit too long, but I see no alternatives to `Activatable` term

Or we can shorten `Parameter` to `Param`, but user can do that as well, so not a lot of sense in here.

Then we have Host Groups to think about.

IMO we can go with `HG` suffix for `HostGroup`, because full `HostGroup` will be confusing when used with `ParameterGroup`

Then we'll have:

- ParameterHG
- ParameterGroupHG
- ActivatableParameterGroupHG

or as a prefix:

- HGParameter
- HGParameterGroup
- HGActivatableParameterGroup

In these cases `HG` works like "modifier" that it's the same entity, but for HostGroups

---

`Desyncable` term speaks about property of an entity, not it's context.